share/lib/python/neuron/coreneuron.py

```diff
@@ -1,92 +1,200 @@
-gpu = False  # Activate GPU computation.
-file_mode = False  # Run via file transfer mode instead of in-memory transfer
-cell_permute = 0  # 0 no permutation; 1 optimize node adjacency
-# 2 optimize parent node adjacency (only for gpu = True)
-warp_balance = 0  # Number of warps to balance. (0 no balance)
-verbose = 2  # 0 quiet, 1 Error, 2 Info, 3 Debug
-prcellstate = -1  # Output prcellstate information for the gid at t=0
-# and at tstop. -1 means no output
-
-
-def property_check(tstop):
-    """
-    Check type and value in range for the user properties.
-    """
-    global enable, gpu, cell_permute, warp_balance, verbose, prcellstate
-    tstop = float(tstop)  # error if can't be a float
-    enable = bool(int(enable))
-    gpu = bool(gpu)
-    cell_permute = int(cell_permute)
-    if gpu and cell_permute < 1:
-        cell_permute = 1  # set cell_permute to 1 by default for gpu exec
-    assert cell_permute in range(3)
-    assert cell_permute in ([1, 2] if gpu else [0, 1])
-    warp_balance = int(warp_balance)
-    assert warp_balance >= 0
-    verbose = int(verbose)
-    verbose = 0 if verbose < 0 else 3 if verbose > 3 else verbose
-    prcellstate = int(prcellstate)
-
-
-def nrncore_arg(tstop):
+import sys
+
+
+class coreneuron(object):
     """
-    Return str that can be used for pc.nrncore_run(str)
-    based on user properties and current NEURON settings.
-    :param tstop: Simulation time (ms)
+    CoreNEURON configuration values.
 
+    An instance of this class stands in as the `neuron.coreneuron` module. Using
+    a class instead of a module allows getter/setter methods to be used, which
+    lets us ensure its properties have consistent types and values.
+
+    Attributes
+    ----------
+    cell_permute
+    enable
+    file_mode
+    gpu
+    prcellstate
+    verbose
+    warp_balance
+
+    Examples
+    --------
+    >>> from neuron import coreneuron
+    >>> coreneuron.enable = True
+    >>> coreneuron.enable
+    True
     """
-    from neuron import h
 
-    arg = ""
-    property_check(tstop)
-    if not enable:
+    def __init__(self):
+        self._enable = False
+        self._gpu = False
+        self._file_mode = False
+        self._cell_permute = None
+        self._warp_balance = 0
+        self._verbose = 2  # INFO
+        self._prcellstate = -1
+
+    def _valid_cell_permute(self):
+        return {1, 2} if self._gpu else {0, 1}
+
+    def _default_cell_permute(self):
+        return 1 if self._gpu else 0
+
+    @property
+    def enable(self):
+        """Use CoreNEURON when calling ParallelContext.psolve(tstop)."""
+        return self._enable
+
+    @enable.setter
+    def enable(self, value):
+        self._enable = bool(int(value))
+
+    @property
+    def gpu(self):
+        """Activate GPU computation in CoreNEURON."""
+        return self._gpu
+
+    @gpu.setter
+    def gpu(self, value):
+        self._gpu = bool(int(value))
```

Comment:
why this `bool(int())`

Reply:
Mostly just because this is how it was before... But also, this is a bit safer if you pass a silly type, e.g. `bool('off') = True`, `bool(int('off'))` is an error 🤷